NovaCustom Boot Guard

mkopec opened the pull request:
*ref: ncm-1807*

miczyg1 committed:
src/soc/intel/alderlake/bootblock/bootblock.c: add CBnT logging

Log CBnT status if CBnT is enabled.

Upstream-Status: Pending
Change-Id: I3ef5cc73ac6440bf76457952f9352dd446be8082
Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

miczyg1 committed:
src/security/vboot/Makefile.inc: add manifests to RO only

There is no need to have them in RW sections.

Upstream-Status: Pending
Change-Id: I1cd6bc4dfc862313493b9cb45eb058c41fdc514a
Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

miczyg1 committed:
src/security/intel/cbnt/Kconfig: handle TPM startup

CbnT does the TPM startup as a part of the ACM execution. It will
cause an invalid postinit error on TPM startup in bootblock or
verstage. Avoid it by selecting the option to ignore invalid postinit.

Upstream-Status: Pending
Change-Id: Idf5e481faf7aa21a8cd934957c5e676de1ee0638
Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

miczyg1 committed:
src/lib/Kconfig.cbfs_verification: allow TOCTOU safety with CBnT

Trust that CBnT has set up a safe environment, if NO_XIP_EARLY_STAGES is
not set by the platform.

Upstream-Status: Pending
Change-Id: I29fba8c5beb367feffe49e8cdfee995adec9cfba
Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

miczyg1 committed:
src/mainboard/clevo/tgl-u/vboot-rwa.fmd: update layout

Increase RW_FWID and RO_FRID to accomodate larger FW IDs.

Upstream-Status: Pending
Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>
Change-Id: I536b36ec4c1e3c5a0e3912bdc6f04b5a88a12c0c

miczyg1 committed:
src/soc/intel/tigerlake/romstage/romstage.c: log CBnT status

Log CBnT status if CBnT is enabled.

Upstream-Status: Pending
Change-Id: I938461176812431e46281832393d64b9bb6a64d4
Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

miczyg1 committed:
src/security/intel/txt: Add necessary adjustments for CBnT

- Restart APs after BIOS ACM LOCK_CONFIG
- Prepare TXT environment for BIOS ACM LOCK_CONFIG
- Set the TXT platform type if CBNT is used
- Restart APs after clearing secrets
- Invalidate cache and reset system depending on ACM parameters

Upstream-Status: Pending
Change-Id: Ib57b4c7e60e37a8f48dd528bc5b23f0306b730e1
Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>
Signed-off-by: Michał Kopeć <michal.kopec@3mdeb.com>
Co-authored-by: SergiiDmytruk <sergii.dmytruk@3mdeb.com>
Signed-off-by: Michał Kopeć <michal.kopec@3mdeb.com>

miczyg1 committed:
Add support for BtG and TXT on Meteor Lake

Add BtG and TXT FSP configuration parameters and add a defconfig for
V560TNx with CBnT enabled.

Upstream-Status: Pending
Change-Id: I84fffa4e4463a2b52f21ae6e768f732ad9cf956d
Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>
Signed-off-by: Michał Kopeć <michal.kopec@3mdeb.com>

miczyg1 committed:
src/soc/intel/meteorlake/Kconfig: Add TXT lib

This is required for CBnT

Upstream-Status: Pending
Change-Id: I5e8fd3ae48908da9d7316fb0c453025b216f291c
Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

miczyg1 committed:
sb/intel/common/firmware: Always use ME binary directly when not stitching ME image

Upstream-Status: Pending
Change-Id: I76a4313bbc3a84dff51dfd5d3c74ea98d25db7c6
Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>
Signed-off-by: Michał Kopeć <michal.kopec@3mdeb.com>

miczyg1 committed:
src/security/intel/txt/common.c: Fix printing error codes

Fix printing minor CBnT error codes

Upstream-Status: Pending
Change-Id: I5ba0965233783c47e86b8899cf0cab048df14180
Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

mkopec commented:
Works on MTL so far but completely untested on TGL and ADL

mkopec committed:
configs/config.novacustom_v5*: enable CBnT

Enable CBnT with placeholder manifests for NovaCustom laptops. The
manifests are replaced with their actual contents by the BtG
provisioning scripts.

Upstream-Status: Pending
Change-Id: I06c57195e08330aa92cce0e25d9555b8b09a792b
Signed-off-by: Michał Kopeć <michal.kopec@3mdeb.com>

mkopec committed:
configs/config.novacustom_n*_tgl: enable CBnT

Enable CBnT on NovaCustom TGL platforms with sample manifests, which are
replaced by BtG provisioning scripts.

Upstream-Status: Pending
Change-Id: I0f5115a752d899b55fe84b58b669bd3841a79588
Signed-off-by: Michał Kopeć <michal.kopec@3mdeb.com>

mkopec committed:
mb/clevo/adl-p: configure FSP for TXT

Configure the necessary FSP parameters for TXT on Alder Lake

Upstream-Status: Pending
Change-Id: If301749bdc6ddb12129b7c86f1926c8ad4d07a62
Signed-off-by: Michał Kopeć <michal.kopec@3mdeb.com>

mkopec committed:
configs/config.novacustom_n*_adl: enable CBnT

Enable CBnT with sample manifests on NovaCustom ADL platforms. The
sample manifests are replaced with their actual contents by BtG
provisioning scripts.

Upstream-Status: Pending
Change-Id: Ifa79f037f86716aa1275e27bdda0db5f1c9b5ec4
Signed-off-by: Michał Kopeć <michal.kopec@3mdeb.com>

mkopec commented:
ADL and TGL are also done now.

mkopec committed:
configs/config.novacustom_v5*: use Vboot RW-A only

CBnT requires a larger ME region, so use an FMAP with enlarged ME and
shrunken BIOS region with RW-A support only.

Upstream-Status: Pending
Change-Id: Ib88f34c7479f41792b13dc6df6dd934e8e62488a
Signed-off-by: Michał Kopeć <michal.kopec@3mdeb.com>

mkopec committed:
configs/config.novacustom_v560tnx_txt: remove redundant config

CBnT has been integrated into the default defconfig, so remove the
redundant TXT defconfig.

Upstream-Status: Pending
Change-Id: Iba0f6952781b85d3cdc2aaf9f6309be68160cff7
Signed-off-by: Michał Kopeć <michal.kopec@3mdeb.com>

mkopec committed:
security/intel/txt/Makefile.mk: Allow build without BIOS ACM

Allow building from source without providing BIOS ACM at build time.

Upstream-Status: Pending
Change-Id: I249a44816e813ba7c4fc1349ba0532deee04f5b3
Signed-off-by: Michał Kopeć <michal.kopec@3mdeb.com>

mkopec committed:
security/intel/cbnt/logging.c: correctly report TPM startup locality

The bitfield values are: 0 for locality 3, 1 for locality 0.

Upstream-Status: Pending
Change-Id: I5ac9671d978af70a8a547e63edc2fe8ff894a371
Signed-off-by: Michał Kopeć <michal.kopec@3mdeb.com>

mkopec committed:
configs/config.novacustom_*: build without ACMs

Allow building from source by not requiring ACMs at build time. ACMs are
added at provisioning time by the provisioning scripts.

Upstream-Status: Pending
Change-Id: Iacc3a1609b4baf051168b4a5b623802117a838d5
Signed-off-by: Michał Kopeć <michal.kopec@3mdeb.com>

mkopec committed:
configs/config.novacustom_*: use public CBnT manifest samples

Use public CBnT manifests that have been uploaded to dasharo-blobs.

Upstream-Status: Inappropriate [Dasharo downstream]
Change-Id: Ia005e36fd16c09784f38abdc3c7a5415fbe8f380
Signed-off-by: Michał Kopeć <michal.kopec@3mdeb.com>

filipleple commented:
There's two failing builds, one with some FSP params mismatch the other can't find a descriptor blob - I think something might've went wrong with the `dasharo-blobs` update. Also if this is the final version I think the earlier commit messages need to be improved to match our current requirements ( esp. `MTL BTG & TXT WIP`)

SergiiDmytruk commented on src/security/intel/txt/getsec_enteraccs.S:
```diff
@@ -109,6 +109,22 @@ cond_push_var_mtrrs:
+	je	invalidate_cache
+	cmp	$5, %eax		/* CLEAR_SECRETS? */
+	je	invalidate_cache
+	cmp	$7, %eax		/* ACHECK? */
```
Can we use constants from `txt_register.h` here instead of hard-coding the values and commenting their meaning?

mkopec replied:
txt_register.h includes src/include/types.h which can't be included in asm, it causes a lot of errors while building:

```
In file included from src/include/types.h:10,
                 from src/security/intel/txt/txt_register.h:6,
                 from src/security/intel/txt/getsec_enteraccs.S:7:
src/include/stdbool.h:6:5: error: "__STDC_VERSION__" is not defined, evaluates to 0 [-Werror=undef]
    6 | #if __STDC_VERSION__ <= 201710L
      |     ^~~~~~~~~~~~~~~~
In file included from src/include/types.h:12:
src/include/stddef.h:22:5: error: "__STDC_VERSION__" is not defined, evaluates to 0 [-Werror=undef]
   22 | #if __STDC_VERSION__ >= 202300L
      |     ^~~~~~~~~~~~~~~~
cc1: all warnings being treated as errors
src/include/stdint.h: Assembler messages:
src/include/stdint.h:7: Error: no such instruction: `typedef signed char int8_t'
src/include/stdint.h:8: Error: no such instruction: `typedef unsigned char uint8_t'
src/include/stdint.h:10: Error: no such instruction: `typedef signed short int16_t'
src/include/stdint.h:11: Error: no such instruction: `typedef unsigned short uint16_t'
src/include/stdint.h:13: Error: no such instruction: `typedef signed int int32_t'
src/include/stdint.h:14: Error: no such instruction: `typedef unsigned int uint32_t'
src/include/stdint.h:16: Error: no such instruction: `typedef signed long long int64_t'
src/include/stdint.h:17: Error: no such instruction: `typedef unsigned long long uint64_t'
src/include/stdint.h:20: Error: no such instruction: `typedef signed long intptr_t'
src/include/stdint.h:21: Error: no such instruction: `typedef unsigned long uintptr_t'
src/include/stdint.h:24: Error: invalid character '(' in mnemonic
src/include/stdint.h:25: Error: invalid character '(' in mnemonic
src/include/stdint.h:27: Error: invalid character '(' in mnemonic
src/include/stdint.h:28: Error: invalid character '(' in mnemonic
src/include/stdint.h:30: Error: invalid character '(' in mnemonic
src/include/stdint.h:31: Error: invalid character '(' in mnemonic
src/include/stdint.h:33: Error: invalid character '(' in mnemonic
src/include/stdint.h:34: Error: invalid character '(' in mnemonic
src/include/stdint.h:36: Error: invalid character '(' in mnemonic
src/include/stdint.h:37: Error: invalid character '(' in mnemonic
src/include/stdint.h:40: Error: no such instruction: `typedef int64_t intmax_t'
src/include/stdint.h:41: Error: no such instruction: `typedef uint64_t uintmax_t'
src/include/stdint.h:44: Error: no such instruction: `typedef int8_t s8'
src/include/stdint.h:45: Error: no such instruction: `typedef uint8_t u8'
src/include/stdint.h:47: Error: no such instruction: `typedef int16_t s16'
src/include/stdint.h:48: Error: no such instruction: `typedef uint16_t u16'
src/include/stdint.h:50: Error: no such instruction: `typedef int32_t s32'
src/include/stdint.h:51: Error: no such instruction: `typedef uint32_t u32'
src/include/stdint.h:53: Error: no such instruction: `typedef int64_t s64'
src/include/stdint.h:54: Error: no such instruction: `typedef uint64_t u64'
src/commonlib/bsd/include/commonlib/bsd/cb_err.h:15: Error: no such instruction: `enum cb_err {'
src/commonlib/bsd/include/commonlib/bsd/cb_err.h:16: Error: junk at end of line, first unrecognized character is `,'
src/commonlib/bsd/include/commonlib/bsd/cb_err.h:17: Error: junk at end of line, first unrecognized character is `,'
src/commonlib/bsd/include/commonlib/bsd/cb_err.h:18: Error: junk at end of line, first unrecognized character is `,'
src/commonlib/bsd/include/commonlib/bsd/cb_err.h:19: Error: junk at end of line, first unrecognized character is `,'
src/commonlib/bsd/include/commonlib/bsd/cb_err.h:22: Error: junk at end of line, first unrecognized character is `,'
src/commonlib/bsd/include/commonlib/bsd/cb_err.h:23: Error: junk at end of line, first unrecognized character is `,'
src/commonlib/bsd/include/commonlib/bsd/cb_err.h:24: Error: junk at end of line, first unrecognized character is `,'
src/commonlib/bsd/include/commonlib/bsd/cb_err.h:25: Error: junk at end of line, first unrecognized character is `,'
src/commonlib/bsd/include/commonlib/bsd/cb_err.h:26: Error: junk at end of line, first unrecognized character is `,'
src/commonlib/bsd/include/commonlib/bsd/cb_err.h:29: Error: junk at end of line, first unrecognized character is `,'
src/commonlib/bsd/include/commonlib/bsd/cb_err.h:30: Error: junk at end of line, first unrecognized character is `,'
src/commonlib/bsd/include/commonlib/bsd/cb_err.h:33: Error: junk at end of line, first unrecognized character is `,'
src/commonlib/bsd/include/commonlib/bsd/cb_err.h:34: Error: junk at end of line, first unrecognized character is `,'
src/commonlib/bsd/include/commonlib/bsd/cb_err.h:35: Error: junk at end of line, first unrecognized character is `,'
src/commonlib/bsd/include/commonlib/bsd/cb_err.h:37: Error: junk at end of line, first unrecognized character is `,'
src/commonlib/bsd/include/commonlib/bsd/cb_err.h:40: Error: junk at end of line, first unrecognized character is `,'
src/commonlib/bsd/include/commonlib/bsd/cb_err.h:41: Error: junk at end of line, first unrecognized character is `,'
src/commonlib/bsd/include/commonlib/bsd/cb_err.h:42: Error: junk at end of line, first unrecognized character is `,'
src/commonlib/bsd/include/commonlib/bsd/cb_err.h:43: Error: junk at end of line, first unrecognized character is `,'
src/commonlib/bsd/include/commonlib/bsd/cb_err.h:46: Error: junk at end of line, first unrecognized character is `,'
src/commonlib/bsd/include/commonlib/bsd/cb_err.h:47: Error: junk at end of line, first unrecognized character is `,'
src/commonlib/bsd/include/commonlib/bsd/cb_err.h:48: Error: junk at end of line, first unrecognized character is `,'
src/commonlib/bsd/include/commonlib/bsd/cb_err.h:49: Error: junk at end of line, first unrecognized character is `,'
src/commonlib/bsd/include/commonlib/bsd/cb_err.h:50: Error: junk at end of line, first unrecognized character is `,'
src/commonlib/bsd/include/commonlib/bsd/cb_err.h:51: Error: junk at end of line, first unrecognized character is `,'
src/commonlib/bsd/include/commonlib/bsd/cb_err.h:52: Error: junk at end of line, first unrecognized character is `,'
src/commonlib/bsd/include/commonlib/bsd/cb_err.h:53: Error: junk at end of line, first unrecognized character is `,'
src/commonlib/bsd/include/commonlib/bsd/cb_err.h:54: Error: junk at end of line, first unrecognized character is `}'
src/include/stdbool.h:7: Error: no such instruction: `typedef _Bool bool'
src/include/stddef.h:8: Error: no such instruction: `typedef long int ptrdiff_t'
src/include/stddef.h:9: Error: no such instruction: `typedef long unsigned int size_t'
src/include/stddef.h:16: Error: no such instruction: `typedef long signed int ssize_t'
src/include/stddef.h:19: Error: no such instruction: `typedef int wchar_t'
src/include/stddef.h:20: Error: no such instruction: `typedef unsigned int wint_t'
src/include/stddef.h:47: Error: no such instruction: `extern char zeroptr[]'
src/security/intel/txt/txt_register.h:174: Error: no such instruction: `union dpr_register {'
src/security/intel/txt/txt_register.h:175: Error: no such instruction: `struct {'
src/security/intel/txt/txt_register.h:176: Error: no such instruction: `uint32_t lock:1'
src/security/intel/txt/txt_register.h:177: Error: no such instruction: `uint32_t prs:1'
src/security/intel/txt/txt_register.h:178: Error: no such instruction: `uint32_t epm:1'
src/security/intel/txt/txt_register.h:179: Error: junk at end of line, first unrecognized character is `1'
src/security/intel/txt/txt_register.h:180: Error: no such instruction: `uint32_t size:8'
src/security/intel/txt/txt_register.h:181: Error: junk at end of line, first unrecognized character is `8'
src/security/intel/txt/txt_register.h:182: Error: no such instruction: `uint32_t top:12'
src/security/intel/txt/txt_register.h:183: Error: junk at end of line, first unrecognized character is `}'
src/security/intel/txt/txt_register.h:184: Error: no such instruction: `uint32_t raw'
src/security/intel/txt/txt_register.h:185: Error: junk at end of line, first unrecognized character is `}'
src/security/intel/txt/txt_register.h:187: Error: no such instruction: `typedef enum {'
src/security/intel/txt/txt_register.h:188: Error: junk at end of line, first unrecognized character is `,'
src/security/intel/txt/txt_register.h:189: Error: junk at end of line, first unrecognized character is `}'
src/security/intel/txt/txt_register.h:191: Error: no such instruction: `typedef enum {'
src/security/intel/txt/txt_register.h:192: Error: junk at end of line, first unrecognized character is `,'
src/security/intel/txt/txt_register.h:193: Error: junk at end of line, first unrecognized character is `,'
src/security/intel/txt/txt_register.h:194: Error: junk at end of line, first unrecognized character is `,'
src/security/intel/txt/txt_register.h:195: Error: junk at end of line, first unrecognized character is `}'
src/security/intel/txt/txt_register.h:202: Error: no such instruction: `struct __attribute__((__packed__))acm_header_v0 {'
src/security/intel/txt/txt_register.h:203: Error: no such instruction: `uint16_t module_type'
src/security/intel/txt/txt_register.h:204: Error: no such instruction: `uint16_t module_sub_type'
src/security/intel/txt/txt_register.h:205: Error: no such instruction: `uint32_t header_len'
src/security/intel/txt/txt_register.h:206: Error: no such instruction: `uint16_t header_version[2]'
src/security/intel/txt/txt_register.h:207: Error: no such instruction: `uint16_t chipset_id'
src/security/intel/txt/txt_register.h:208: Error: no such instruction: `uint16_t flags'
src/security/intel/txt/txt_register.h:209: Error: no such instruction: `uint32_t module_vendor'
src/security/intel/txt/txt_register.h:210: Error: no such instruction: `uint32_t date'
src/security/intel/txt/txt_register.h:211: Error: no such instruction: `uint32_t size'
src/security/intel/txt/txt_register.h:212: Error: no such instruction: `uint16_t txt_svn'
src/security/intel/txt/txt_register.h:213: Error: no such instruction: `uint16_t se_svn'
src/security/intel/txt/txt_register.h:214: Error: no such instruction: `uint32_t code_control'
src/security/intel/txt/txt_register.h:215: Error: no such instruction: `uint32_t error_entry_point'
src/security/intel/txt/txt_register.h:216: Error: no such instruction: `uint32_t gdt_limit'
src/security/intel/txt/txt_register.h:217: Error: no such instruction: `uint32_t gdt_ptr'
src/security/intel/txt/txt_register.h:218: Error: no such instruction: `uint32_t seg_sel'
src/security/intel/txt/txt_register.h:219: Error: no such instruction: `uint32_t entry_point'
src/security/intel/txt/txt_register.h:220: Error: no such instruction: `uint8_t reserved2[64]'
src/security/intel/txt/txt_register.h:221: Error: no such instruction: `uint32_t key_size'
src/security/intel/txt/txt_register.h:222: Error: no such instruction: `uint32_t scratch_size'
src/security/intel/txt/txt_register.h:223: Error: no such instruction: `uint8_t rsa2048_pubkey[256]'
src/security/intel/txt/txt_register.h:224: Error: no such instruction: `uint32_t pub_exp'
src/security/intel/txt/txt_register.h:225: Error: no such instruction: `uint8_t rsa2048_sig[256]'
src/security/intel/txt/txt_register.h:226: Error: no such instruction: `uint32_t scratch[143]'
src/security/intel/txt/txt_register.h:227: Error: no such instruction: `uint8_t user_area[]'
src/security/intel/txt/txt_register.h:228: Error: junk at end of line, first unrecognized character is `}'
src/security/intel/txt/txt_register.h:230: Error: no such instruction: `struct __attribute__((__packed__))acm_info_table {'
src/security/intel/txt/txt_register.h:231: Error: no such instruction: `uint8_t uuid[16]'
src/security/intel/txt/txt_register.h:232: Error: no such instruction: `uint8_t chipset_acm_type'
src/security/intel/txt/txt_register.h:233: Error: no such instruction: `uint8_t version'
src/security/intel/txt/txt_register.h:234: Error: no such instruction: `uint16_t length'
src/security/intel/txt/txt_register.h:235: Error: no such instruction: `uint32_t chipset_id_list'
src/security/intel/txt/txt_register.h:236: Error: no such instruction: `uint32_t os_sinit_data_ver'
src/security/intel/txt/txt_register.h:237: Error: no such instruction: `uint32_t min_mle_hdr_ver'
src/security/intel/txt/txt_register.h:238: Error: no such instruction: `uint32_t capabilities'
src/security/intel/txt/txt_register.h:239: Error: no such instruction: `uint8_t acm_ver'
src/security/intel/txt/txt_register.h:240: Error: no such instruction: `uint8_t reserved[3]'
src/security/intel/txt/txt_register.h:241: Error: junk at end of line, first unrecognized character is `}'
src/security/intel/txt/txt_register.h:248: Error: no such instruction: `struct __attribute__((__packed__))txt_extended_data_element_header {'
src/security/intel/txt/txt_register.h:249: Error: no such instruction: `uint32_t type'
src/security/intel/txt/txt_register.h:250: Error: no such instruction: `uint32_t size'
src/security/intel/txt/txt_register.h:251: Error: no such instruction: `uint8_t data[0]'
src/security/intel/txt/txt_register.h:252: Error: junk at end of line, first unrecognized character is `}'
src/security/intel/txt/txt_register.h:259: Error: no such instruction: `struct __attribute__((__packed__))txt_bios_spec_ver_element {'
src/security/intel/txt/txt_register.h:260: Error: no such instruction: `struct txt_extended_data_element_header header'
src/security/intel/txt/txt_register.h:261: Error: no such instruction: `uint16_t ver_major'
src/security/intel/txt/txt_register.h:262: Error: no such instruction: `uint16_t ver_minor'
src/security/intel/txt/txt_register.h:263: Error: no such instruction: `uint16_t ver_revision'
src/security/intel/txt/txt_register.h:264: Error: junk at end of line, first unrecognized character is `}'
src/security/intel/txt/txt_register.h:267: Error: no such instruction: `struct __attribute__((__packed__))txt_heap_acm_element1 {'
src/security/intel/txt/txt_register.h:268: Error: no such instruction: `struct txt_extended_data_element_header header'
src/security/intel/txt/txt_register.h:269: Error: no such instruction: `uint32_t num_acms'
src/security/intel/txt/txt_register.h:270: Error: no such instruction: `uint64_t acm_addrs[1]'
src/security/intel/txt/txt_register.h:271: Error: junk at end of line, first unrecognized character is `}'
src/security/intel/txt/txt_register.h:274: Error: no such instruction: `struct __attribute__((__packed__))txt_heap_acm_element2 {'
src/security/intel/txt/txt_register.h:275: Error: no such instruction: `struct txt_extended_data_element_header header'
src/security/intel/txt/txt_register.h:276: Error: no such instruction: `uint32_t num_acms'
src/security/intel/txt/txt_register.h:277: Error: no such instruction: `uint64_t acm_addrs[2]'
src/security/intel/txt/txt_register.h:278: Error: junk at end of line, first unrecognized character is `}'
src/security/intel/txt/txt_register.h:285: Error: no such instruction: `struct __attribute__((__packed__))txt_biosdataregion {'
src/security/intel/txt/txt_register.h:286: Error: no such instruction: `uint32_t version'
src/security/intel/txt/txt_register.h:287: Error: no such instruction: `uint32_t bios_sinit_size'
src/security/intel/txt/txt_register.h:288: Error: no such instruction: `uint64_t lcp_pd_base'
src/security/intel/txt/txt_register.h:289: Error: no such instruction: `uint64_t lcp_pd_size'
src/security/intel/txt/txt_register.h:290: Error: no such instruction: `uint32_t no_logical_procs'
src/security/intel/txt/txt_register.h:291: Error: no such instruction: `uint32_t sinit_flags'
src/security/intel/txt/txt_register.h:292: Error: no such instruction: `union {'
src/security/intel/txt/txt_register.h:293: Error: no such instruction: `uint32_t mle_flags'
src/security/intel/txt/txt_register.h:294: Error: no such instruction: `struct {'
src/security/intel/txt/txt_register.h:295: Error: no such instruction: `uint32_t support_acpi_ppi:1'
src/security/intel/txt/txt_register.h:296: Error: no such instruction: `uint32_t platform_type:2'
src/security/intel/txt/txt_register.h:297: Error: junk at end of line, first unrecognized character is `}'
src/security/intel/txt/txt_register.h:298: Error: junk at end of line, first unrecognized character is `}'
src/security/intel/txt/txt_register.h:299: Error: no such instruction: `u8 extended_data_elements[0]'
src/security/intel/txt/txt_register.h:300: Error: junk at end of line, first unrecognized character is `}'
src/security/intel/txt/txt_register.h:302: Error: no such instruction: `void txt_dump_regions(void)'
src/security/intel/txt/txt_register.h:303: Error: no such instruction: `void txt_dump_chipset_info(void)'
src/security/intel/txt/txt_register.h:304: Error: no such instruction: `void txt_dump_acm_info(const struct acm_header_v0 *acm_header)'
src/security/intel/txt/txt_register.h:305: Error: no such instruction: `void txt_dump_getsec_parameters(void)'
make: *** [Makefile:430: build/ramstage/security/intel/txt/getsec_enteraccs.o] Error 1
```

miczyg1 replied:
Try `#ifndef __ASSEMBLER__` to guard everything that is C and not not pre-processor

SergiiDmytruk replied:
> Try `#ifndef __ASSEMBLER__` to guard everything that is C and not not pre-processor

Probably not everything, just the top part before the first `struct` will be enough (`#include` needs to be moved right before it).  There are some `#define`s interleaved with structures, but we won't use them.

mkopec replied:
Right, done: https://github.com/Dasharo/coreboot/pull/685/commits/55d7b77e4ebec3b635f871c685d40d557754cfb8

SergiiDmytruk commented on src/security/intel/txt/getsec_enteraccs.S:
```diff
@@ -213,9 +229,29 @@ cond_clear_var_mtrrs:
 
 	/* Enable SMX and FXSTORE - for getsec */
```
Either:
```suggestion
	/* Enable SMX and FXSTORE - for getsec; make accesses of DR4/DR5 cause #UD (invalid opcode) */
```
Or maybe we don't need to set `CR4_DE`?  Not sure what could require it.

mkopec replied:
Squashed to https://github.com/Dasharo/coreboot/pull/685/commits/8df8e284be6c4e362c6e026c31de13196312d5ab , I'll defer to @miczyg1 on whether we need to set CR4_DE

miczyg1 replied:
I mimicked Intel's reference implementation which always set DE in CR4. Probably not required. Maybe worth testing.

SergiiDmytruk commented on src/security/intel/txt/getsec_enteraccs.S:
```diff
@@ -213,9 +229,29 @@ cond_clear_var_mtrrs:
+
+#if CONFIG(INTEL_CBNT_SUPPORT)
+	/* Clean all MCi_STATUS MSR registers */
+	movl	$0x179, %ecx		/* MCG_CAP */
```
`IA32_MCG_CAP` in`include/cpu/x86/msr.h`.

mkopec replied:
Done and squashed to https://github.com/Dasharo/coreboot/pull/685/commits/8df8e284be6c4e362c6e026c31de13196312d5ab

SergiiDmytruk replied:
The comment can be dropped now.

mkopec replied:
https://github.com/Dasharo/coreboot/pull/685/commits/660dbaa6977ae85f33bda8a6727b2c01c3f23127

SergiiDmytruk commented on src/security/intel/txt/getsec_enteraccs.S:
```diff
@@ -213,9 +229,29 @@ cond_clear_var_mtrrs:
+	movzx	%al, %ebx		/* ebx = MCR bank count */
+	xorl	%eax, %eax		/* Write 0 into all MCi_STATUS registers */
+	xorl	%edx, %edx
+	movl	$0x401, %ecx		/* MC0_STATUS */
```
`IA32_MC0_STATUS` in `include/cpu/x86/msr.h`.

mkopec replied:
Done and squashed to https://github.com/Dasharo/coreboot/pull/685/commits/8df8e284be6c4e362c6e026c31de13196312d5ab

SergiiDmytruk replied:
a/a, the comment is unnecessary now.

mkopec replied:
https://github.com/Dasharo/coreboot/pull/685/commits/660dbaa6977ae85f33bda8a6727b2c01c3f23127

SergiiDmytruk commented on src/security/intel/txt/getsec_enteraccs.S:
```diff
@@ -269,6 +305,15 @@ cond_clear_var_mtrrs:
+	cmp	$5, %eax 		/* CLEAR_SECRETS? */
+	je	reset_system
+	cmp	$7, %eax		/* ACHECK? */
+	je	reset_system
```
Another place for constants from `txt_register.h`.

mkopec replied:
a/a, can't include that header due to types.h not working with asm files

mkopec replied:
https://github.com/Dasharo/coreboot/pull/685/commits/55d7b77e4ebec3b635f871c685d40d557754cfb8

SergiiDmytruk commented on src/security/intel/txt/getsec_enteraccs.S:
```diff
@@ -334,3 +379,17 @@ cond_pop_var_mtrrs:
+	outb	%al, %dx
+
+	movw	$0xcf9, %dx
+	movb	$0x0e, %al
```
There are constants in `arch/x86/include/cf9_reset.h`, but a `/* FULL_RST | RST_CPU | SYS_RST */` comment will probably suffice.

mkopec replied:
Done and squashed to https://github.com/Dasharo/coreboot/pull/685/commits/8df8e284be6c4e362c6e026c31de13196312d5ab

SergiiDmytruk commented on src/security/intel/txt/ramstage.c:
```diff
@@ -202,6 +222,13 @@ static void push_sinit_heap(u8 **heap_ptr, void *data, size_t data_length)
+{
+	msr_t msr = rdmsr(MSR_BOOT_GUARD_SACM_INFO);
+
+	return !!(msr.hi & (B_BOOT_GUARD_SACM_INFO_TXT_CAPABILITY >> 32));
```
```suggestion
	return !!(msr.raw & B_BOOT_GUARD_SACM_INFO_TXT_CAPABILITY);
```
I went to check constants and types because of the way this was written.

mkopec replied:
Applied and squashed to https://github.com/Dasharo/coreboot/pull/685/commits/db665ed3b6469c3fbad3743c3c0752f4324f36ac

SergiiDmytruk commented on src/security/intel/txt/ramstage.c:
```diff
@@ -222,7 +249,16 @@ static void txt_heap_fill_common_bdr(struct txt_biosdataregion *bdr)
+			bdr->platform_type = 1;
+		}
+	} else {
+		bdr->platform_type = 0;
```
Worth comments on assigning platform types:
* 0: legacy state / platform undefined
* 1: client platform, client SINIT is required
* 2: server platform, server SINIT is required

mkopec replied:
Updated & squashed to the original commit: https://github.com/Dasharo/coreboot/pull/685/commits/714525fdc8a13ea2d9c7ba97c68470eccb623424

SergiiDmytruk replied:
nit: `auto-detect` instead of `auto-decect`

mkopec replied:
derp, thanks for noticing: https://github.com/Dasharo/coreboot/pull/685/commits/b193d41f81e40f1e7cac69a9023c2a029caac0b1

SergiiDmytruk commented on src/southbridge/intel/common/firmware/Makefile.mk:
```diff
@@ -30,6 +30,10 @@ OBJ_ME_BIN := $(obj)/me.bin
+OBJ_ME_BIN := $(call strip_quotes,$(CONFIG_ME_BIN_PATH))
+
+else
+
```
Can `OBJ_ME_BIN := $(obj)/me.bin` line from above be moved here?  `Always use ME binary directly` seemingly contradicts the check of `CONFIG_STITCH_ME_BIN`, I think it was meant to say `Always use ME binary directly when not stitching ME image` or something to that effect.

mkopec replied:
Squashed to https://github.com/Dasharo/coreboot/pull/685/commits/96ad51c075ffa151559dde42e112918e37427b50

SergiiDmytruk replied:
The new version looks quite different from e77596d, this part is gone:
```make
$(OBJ_ME_BIN): $(call strip_quotes,$(CONFIG_ME_BIN_PATH))
	cp $< $@
```

mkopec replied:
Added back: https://github.com/Dasharo/coreboot/pull/685/commits/4ba00ef515fabdfe80b77d5ccc95f9e490eacf01

SergiiDmytruk commented on src/security/intel/cbnt/Makefile.mk:
```diff
@@ -102,10 +102,12 @@ endif # CONFIG_INTEL_CBNT_BPM_ONLY_UNSIGNED
 
 ifneq ($(CONFIG_INTEL_CBNT_BOOT_POLICY_MANIFEST_BINARY),"")
+
+#regions-for-file-boot_policy_manifest.bin = MANIFESTS
```
This comment shouldn't stay.

mkopec replied:
removed and squashed to https://github.com/Dasharo/coreboot/pull/685/commits/b0b72ca24cf7397e4d05953c1abee66faef1d5fd

SergiiDmytruk commented on src/security/intel/cbnt/Makefile.mk:
```diff
@@ -152,10 +154,11 @@ show_notices::
 	@printf "$$ ifittool -r COREBOOT -a -n key_manifest.bin -t 11 -s 12 -f build/coreboot.rom\n"
 
 else
+# regions-for-file-key_manifest.bin = MANIFESTS
```
Should also be removed.

mkopec replied:
removed and squashed to https://github.com/Dasharo/coreboot/pull/685/commits/b0b72ca24cf7397e4d05953c1abee66faef1d5fd

SergiiDmytruk commented on src/mainboard/clevo/adl-p/variants/nv40pz/romstage.c:
```diff
@@ -21,5 +25,39 @@ void mainboard_memory_init_params(FSPM_UPD *mupd)
+	size_t acm_size = 0;
+	uintptr_t acm_base;
+
+	if (CONFIG(INTEL_TXT)) {
```
This `if` can be dropped.

mkopec replied:
https://github.com/Dasharo/coreboot/pull/685/commits/507f28e0b9ba3f8225b7d9b18150effd3f76065d

SergiiDmytruk commented on src/mainboard/clevo/adl-p/variants/ns50pu/romstage.c:
```diff
@@ -27,5 +31,38 @@ void mainboard_memory_init_params(FSPM_UPD *mupd)
 
 	mupd->FspmConfig.CpuPcieRpClockReqMsgEnable[0] = 1;
 
+#if CONFIG(INTEL_TXT)
```
nit: lack of `/* Use pre-processor because CONFIG_INTEL_TXT_CBFS_BIOS_ACM is not defined otherwise */` here is inconsistent

mkopec replied:
https://github.com/Dasharo/coreboot/pull/685/commits/507f28e0b9ba3f8225b7d9b18150effd3f76065d

SergiiDmytruk reviewed (changes requested):
In addition to comments commits need rebasing to clean up changes and better messages at times (especially `MTL BTG & TXT WIP`).

Should this be still marked as a draft?

SergiiDmytruk commented on configs/config.novacustom_ns5x_adl:
```diff
@@ -27,6 +27,15 @@ CONFIG_DRIVERS_EFI_MAIN_FW_LSV=0x01080002
+# CONFIG_INTEL_CBNT_GENERATE_KM is not set
+# CONFIG_INTEL_CBNT_GENERATE_BPM is not set
+CONFIG_INTEL_CBNT_KEY_MANIFEST_BINARY="3rdparty/dasharo-blobs/cbnt/km_sample.bin"
+CONFIG_INTEL_CBNT_BOOT_POLICY_MANIFEST_BINARY="3rdparty/dasharo-blobs/cbnt/km_sample.bin"
```
```suggestion
CONFIG_INTEL_CBNT_BOOT_POLICY_MANIFEST_BINARY="3rdparty/dasharo-blobs/cbnt/bpm_sample.bin"
```

SergiiDmytruk commented on configs/config.novacustom_ns5x_tgl:
```diff
@@ -27,6 +27,15 @@ CONFIG_DRIVERS_EFI_MAIN_FW_LSV=0x01060002
+# CONFIG_INTEL_CBNT_GENERATE_KM is not set
+# CONFIG_INTEL_CBNT_GENERATE_BPM is not set
+CONFIG_INTEL_CBNT_KEY_MANIFEST_BINARY="3rdparty/dasharo-blobs/cbnt/km_sample.bin"
+CONFIG_INTEL_CBNT_BOOT_POLICY_MANIFEST_BINARY="3rdparty/dasharo-blobs/cbnt/km_sample.bin"
```
```suggestion
CONFIG_INTEL_CBNT_BOOT_POLICY_MANIFEST_BINARY="3rdparty/dasharo-blobs/cbnt/bpm_sample.bin"
```

SergiiDmytruk commented on configs/config.novacustom_nv4x_adl:
```diff
@@ -28,6 +28,15 @@ CONFIG_DRIVERS_EFI_MAIN_FW_LSV=0x01080002
+# CONFIG_INTEL_CBNT_GENERATE_KM is not set
+# CONFIG_INTEL_CBNT_GENERATE_BPM is not set
+CONFIG_INTEL_CBNT_KEY_MANIFEST_BINARY="3rdparty/dasharo-blobs/cbnt/km_sample.bin"
+CONFIG_INTEL_CBNT_BOOT_POLICY_MANIFEST_BINARY="3rdparty/dasharo-blobs/cbnt/km_sample.bin"
```
```suggestion
CONFIG_INTEL_CBNT_BOOT_POLICY_MANIFEST_BINARY="3rdparty/dasharo-blobs/cbnt/bpm_sample.bin"
```

SergiiDmytruk commented on configs/config.novacustom_nv4x_tgl:
```diff
@@ -27,6 +27,15 @@ CONFIG_DRIVERS_EFI_MAIN_FW_LSV=0x01060002
+# CONFIG_INTEL_CBNT_GENERATE_KM is not set
+# CONFIG_INTEL_CBNT_GENERATE_BPM is not set
+CONFIG_INTEL_CBNT_KEY_MANIFEST_BINARY="3rdparty/dasharo-blobs/cbnt/km_sample.bin"
+CONFIG_INTEL_CBNT_BOOT_POLICY_MANIFEST_BINARY="3rdparty/dasharo-blobs/cbnt/km_sample.bin"
```
```suggestion
CONFIG_INTEL_CBNT_BOOT_POLICY_MANIFEST_BINARY="3rdparty/dasharo-blobs/cbnt/bpm_sample.bin"
```

SergiiDmytruk reviewed (changes requested):
`km_sample.bin` is used instead of `bpm_sample.bin` in multiple places.  Not that it makes huge difference, but since we have 2 files, should use both of them.

P.S. `_sample` suffix is a misnomer, these files are placeholders.

mkopec committed:
mb/clevo/adl-p: fix up usage of CONFIG(INTEL_TXT) for guarding CBnT

- Add missing comment to Clevo NS50PU board
- Don't double-check CONFIG(INTEL_TXT) if it's already guarded by the
  preprocessor

Upstream-Status: Pending
Change-Id: If5b99655357f055a1b204d1c02c220f9b6a1fc7d
Signed-off-by: Michał Kopeć <michal.kopec@3mdeb.com>

mkopec committed:
configs/novacustom_n*: use bpm_sample as the boot policy manifest

Originally km_sample was used as the BPM due to a copy-paste error.

Upstream-Status: Pending
Change-Id: Icc44520ba9da62b15ad1c45c4ed25e98de523412
Signed-off-by: Michał Kopeć <michal.kopec@3mdeb.com>

mkopec commented:
@SergiiDmytruk @filipleple I've updated all the commit messages. Will try to resolve CI build issues now

mkopec commented:
CI seems happy now.

mkopec committed:
configs: update dasharo edk2 for CBnT status and enable it on NovaCustom

Enable CBnT status page on all novacustom laptops, keep it hidden on
other platforms.

Upstream-Status: Inappropriate [Dasharo downstream]
Change-Id: I6665532a6c69622e4e8436d9e7fb689534759569
Signed-off-by: Michał Kopeć <michal.kopec@3mdeb.com>

mkopec committed:
configs: update dasharo edk2 for CBnT status page

Upstream-Status: Inappropriate [Dasharo downstream]
Change-Id: I6665532a6c69622e4e8436d9e7fb689534759569
Signed-off-by: Michał Kopeć <michal.kopec@3mdeb.com>

mkopec committed:
payloads/edk2: add option to control CBnT menu visibility

Upstream-Status: Inappropriate [Dasharo downstream]
Change-Id: Ia3d19e1b745396bc81c5d0888a73810c784ce4a2
Signed-off-by: Michał Kopeć <michal.kopec@3mdeb.com>

mkopec committed:
payloads/edk2/Kconfig.dasharo: make btg status depend on cbnt and ME page

Upstream-Status: Inappropriate [Dasharo downstream]
Change-Id: Ie8ff91cfc71f9478c67876aeb71f7b89483fff90
Signed-off-by: Michał Kopeć <michal.kopec@3mdeb.com>

mkopec committed:
security/intel/txt: use defined values instead of hardcoding ACM params

Use existing defines from txt_register.h instead of hand-coding ACM
parameters.

Upstream-Status: Pending
Change-Id: Iea37b504a40404a78455ff80b80c88ea290b16f3
Signed-off-by: Michał Kopeć <michal.kopec@3mdeb.com>

mkopec committed:
configs: bump edk2 for the merged version of cbnt status

Upstream-Status: Inappropriate [Dasharo downstream]
Change-Id: Ifb083b485716e6bdcc05431c2844abadb7278acb
Signed-off-by: Michał Kopeć <michal.kopec@3mdeb.com>

mkopec committed:
security/intel/txt/getsec_enteraccs.S: drop redundant comments

Upstream-Status: Pending
Change-Id: Id46b5bfc9983e61b6a13d6798796e71b14c94dfc
Signed-off-by: Michał Kopeć <michal.kopec@3mdeb.com>

mkopec committed:
security/intel/txt/ramstage.c: fix typo in comment

Upstream-Status: Pending
Change-Id: I8209a95d46acfedcfc3879e666c32d24ddc9a3f3
Signed-off-by: Michał Kopeć <michal.kopec@3mdeb.com>

mkopec commented:
Rebased and resolved conflicts.

mkopec commented on src/mainboard/novacustom/mtl-h/vboot-rwab.fmd:
```diff
@@ -35,16 +35,18 @@ FLASH 32M {
 			RO_VPD(PRESERVE) 16K
 			RO_GSCVD 8K
 			RO_SECTION {
+				MANIFESTS(CBFS) 16K
```
Unused region, unnecesarilly moves FMAP higher.

mkopec replied:
https://github.com/Dasharo/coreboot/pull/685/commits/93cd245884f1c113cfdbb22e087856a5142c7156

mkopec committed:
mb/novacustom/novacustom/mtl-h/vboot-rwab.fmd: drop unused MANIFESTS region

This region is unused and moves the FMAP for no good reason, making
flashing using flashrom with the --fmap option potentially problematic

Upstream-Status: Pending
Change-Id: Ibfa33537c95279f8354ecd04e2afbd90c00ae95b
Signed-off-by: Michał Kopeć <michal.kopec@3mdeb.com>

mkopec committed:
mb/{clevo,novacustom}: Enable Vboot CBFS fallback

This is needed to make sure the BIOS ACM is accessible post-bootblock.

Upstream-Status: Pending
Change-Id: I6676555852d395f3f70721297e7e7ff9af46a1f4
Signed-off-by: Michał Kopeć <michal.kopec@3mdeb.com>

SergiiDmytruk commented on src/soc/intel/meteorlake/fsp_params.c:
```diff
@@ -406,6 +406,9 @@ static void fill_fsps_cpu_params(FSP_S_CONFIG *s_cfg,
 	if (CONFIG(USE_FSP_FEATURE_PROGRAM_ON_APS))
 		fill_fsps_microcode_params(s_cfg, config);
+
+
```
nit: extra empty line (I thought code lint should catch such things)

mkopec replied:
squashed to https://github.com/Dasharo/coreboot/pull/685/commits/85676d6daff2155a5a47ddd79731ab0de3abcf6d

SergiiDmytruk commented on src/security/intel/txt/getsec_enteraccs.S:
```diff
@@ -211,11 +230,31 @@ cond_clear_var_mtrrs:
+	orl	$(CR4_SMXE | CR4_DE | CR4_OSFXSR), %eax
 	movl	%eax, %cr4
 
+
```
nit: extra empty line

mkopec replied:
squashed to https://github.com/Dasharo/coreboot/pull/685/commits/f6c0615ad974e6ddf73233cd67f8e6a82287fa4d

SergiiDmytruk reviewed (dismissed):
A few of the new commits could be squashed into previous ones before merging and there are couple empty lines, but otherwise looks good.

mkopec commented:
@filipleple @SergiiDmytruk I addressed your review and rebased but github removed the approvals, can you take another look?

SergiiDmytruk reviewed (approved):
`Upstream-Status:` needs to be added to make CI happy.
